src/gbinder_fmq.c

```diff
@@ -0,0 +1,782 @@
+        if (base_address) {
+            munmap(base_address, map_length);
+        }
+    }
+}
+
+static
+void
+gbinder_fmq_free(
+    GBinderFmq* self)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    if (self->desc) {
+        if (self->desc->flags == GBINDER_FMQ_TYPE_UNSYNC_WRITE) {
+            g_free(priv->read_ptr);
+        } else {
+            gbinder_fmq_unmap_grantor_descriptor(self, priv->read_ptr,
+                READ_PTR_POS);
+        }
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->write_ptr,
+            WRITE_PTR_POS);
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->ring,
+            DATA_PTR_POS);
+        gbinder_fmq_unmap_grantor_descriptor(self, priv->event_flag_ptr,
+            EVENT_FLAG_PTR_POS);
+
+        g_free((GBinderFmqGrantorDescriptor*)self->desc->grantors.data.ptr);
+        g_free((GBinderFds*)self->desc->data.fds);
+
+        g_free(self->desc);
+    }
+    g_slice_free(GBinderFmqPriv, priv);
+}
+
+/* Public API */
+
+GBinderFmq*
+gbinder_fmq_new(
+    gsize item_size,
+    gsize num_items,
+    GBINDER_FMQ_TYPE type,
+    GBINDER_FMQ_FLAGS flags,
+    gint fd,
+    gsize buffer_size)
+{
+    GBinderFmq* self = NULL;
+
+    if (item_size <= 0) {
+        GWARN("Incorrect item size");
+    } else if (num_items <= 0) {
+        GWARN("Empty queue requested");
+    } else if (num_items > SIZE_MAX / item_size) {
+        GWARN("Requested message queue size too large");
+    } else if (fd != -1 && num_items * item_size > buffer_size) {
+        GWARN("The size needed for items (%"G_GSIZE_FORMAT") is larger\
+            than the supplied buffer size (%"G_GSIZE_FORMAT")",
+            num_items * item_size, buffer_size);
+    } else {
+        gboolean configure_event_flag =
+            (flags & GBINDER_FMQ_FLAG_CONFIGURE_EVENT_FLAG) != 0;
+        gsize queue_size_bytes = num_items * item_size;
+        gsize meta_data_size;
+        gsize shmem_size;
+        int shmem_fd;
+        GBinderFmqPriv* priv = g_slice_new0(GBinderFmqPriv);
+
+        self = &priv->pub;
+
+        meta_data_size = 2 * sizeof(guint64);
+        if (configure_event_flag) {
+            meta_data_size += sizeof(guint32);
+        }
+
+        /* Allocate shared memory */
+        if (fd != -1) {
+            /* User-supplied ringbuffer memory provided,
+             * allocating memory only for meta data */
+            shmem_size = (meta_data_size + getpagesize() - 1) &
+                ~(getpagesize() - 1);
+        } else {
+            /* Allocate ringbuffer, read counter and write counter */
+            shmem_size = (G_ALIGN8(queue_size_bytes) +
+                meta_data_size + getpagesize() - 1) & ~(getpagesize() - 1);
+        }
+
+        shmem_fd = syscall(__NR_memfd_create, "MessageQueue", MFD_CLOEXEC);
+
+        if (shmem_fd < 0 || ftruncate(shmem_fd, shmem_size) < 0) {
+            GWARN("Failed to create shared memory file");
+            gbinder_fmq_free(self);
```

Comment:
and return NULL I guess

Reply:
I now set self to NULL here and return will happen at the end of the function.